### PR TITLE
Verify personal data

### LINKED ISSUE #3127: Select own Lookup Server
In some use-cases, you may not want to use `lookup.nextcloud.com` as your lookup-server for Nextcloud instances. This would be the case if you are running some Nextclouds in a larger envoiroment, which should communicate among each other (e.g. `lookup.bigcompany.com` - but not to all Nextclouds out there) ... 

I know that this is a very small use case currently but if we look into the future of federation - this could be more and more interessting ...running some small Nextclouds connected with a company-controlled lookup-server could replace complex clustered instances in some cases 😁😅 

I think we should make this configurable. Maybe in the `config.php` or directly inside the `sharing`-settings.

What do you think? @schiessle

### PR DESCRIPTION
**Pull request by schiessle**
Allow users to verify their email address, twitter account and webpage for the lookup server

Fix #3127

### CODE REVIEW

File: apps/lookup_server_connector/appinfo/app.php [hunk 1]
```diff
@@ -28,18 +28,20 @@
 		\OC::$server->getAppDataDir('identityproof'),
 		\OC::$server->getCrypto()
 	);
+
+	$config = \OC::$server->getConfig();
+	$lookupServer = $config->getSystemValue('lookup_server', '');
```
Comment (MorrisJobke) [inline apps/lookup_server_connector/appinfo/app.php]: No default value?

File: config/config.sample.php [hunk 2]
```diff
@@ -1491,4 +1491,9 @@
  */
 'copied_sample_config' => true,
 
+/**
+ * use a custom lookup server to publish user data
+ */
+'lookup_server' => 'https://lookup.nextcloud.com/users',
```
Comment (MorrisJobke) [inline config/config.sample.php]: I guess the `/users` needs to be removed here.

File: apps/lookup_server_connector/lib/UpdateLookupServer.php [hunk 3]
```diff
@@ -40,44 +37,36 @@
...
 	private $clientService;
-	/** @var Manager */
-	private $keyManager;
 	/** @var Signer */
 	private $signer;
 	/** @var IJobList */
 	private $jobList;
 	/** @var string URL point to lookup server */
-	private $lookupServer = 'https://lookup.nextcloud.com/users';
+	private $lookupServer = 'https://lookup.nextcloud.com';
```
Comment (MorrisJobke) [inline apps/lookup_server_connector/lib/UpdateLookupServer.php]: Use the default while fetching from the config instead.

File: apps/lookup_server_connector/lib/UpdateLookupServer.php [hunk 4]
```diff
@@ -40,44 +37,36 @@
...
 	 */
 	public function __construct(AccountManager $accountManager,
-								IConfig $config,
-								ISecureRandom $secureRandom,
 								IClientService $clientService,
-								Manager $manager,
 								Signer $signer,
-								IJobList $jobList) {
+								IJobList $jobList,
+								$lookupServer = '') {
```
Comment (nickvergessen) [inline apps/lookup_server_connector/lib/UpdateLookupServer.php]: 👎

File: config/config.sample.php (hunk 2 above)
Comment (schiessle) [inline config/config.sample.php]: no, that's how our current lookup server is configured. That's also how it is in stable11

File: apps/lookup_server_connector/appinfo/app.php (hunk 1 above)
Comment (schiessle) [inline apps/lookup_server_connector/appinfo/app.php]: The default value is that nothing is configured and than the lookup connector will fall back to the default URL.

File: apps/lookup_server_connector/lib/UpdateLookupServer.php (hunk 3 above)
Comment (schiessle) [inline apps/lookup_server_connector/lib/UpdateLookupServer.php]: I prefer to have one place where the default defined in one place instead of spread around in multiple places. This makes it much easier to change it if necessary.

File: apps/lookup_server_connector/lib/UpdateLookupServer.php (hunk 4 above)
Comment (schiessle) [inline apps/lookup_server_connector/lib/UpdateLookupServer.php]: Related to: https://github.com/nextcloud/server/pull/3869#discussion_r106896494
In most cases people will just use the default value. People who use this class shouldn't have to think about it or even know the default server.

**Comment by schiessle** [comment]
Thanks for the early review. But keep in mind that most code will probably change before it is ready to review. That's a early development branch.

**Comment by schiessle** [comment]
@jancborchardt please don't forget the icons we discussed during the hackweek.... Thanks!

File: apps/lookup_server_connector/lib/UpdateLookupServer.php (hunk 3 above)
Comment (MorrisJobke) [inline apps/lookup_server_connector/lib/UpdateLookupServer.php]: The problem I see with this approach: How to set "no lookup server"?

Comment (schiessle) [inline apps/lookup_server_connector/lib/UpdateLookupServer.php]: there is always one configured lookup server. That's also the case for Nc11. You can disable the whole functionality in the admin settings, no need to play around with config.php variables.

**Comment by MorrisJobke** [comment]
@jancborchardt Please fix the layout:

bildschirmfoto 2017-04-07 um 12 51 21

**Comment by schiessle** [comment]
@MorrisJobke for now the text label are just there to allow me to work on the actual functionality, I didn't put a lot of energy in getting the layout right (although I have to say that it looks better on my screen 😉 ).

Still waiting for some nice icons by @jancborchardt. Bonus points for Jan if he adds them directly to this branch with the right layout 😉

**Comment by MorrisJobke** [comment]
@jancborchardt Please have a look at this or this will not land in 12.

**Comment by jancborchardt** [comment]
Sorry, have been drowning in notifications – will check it out.

**Comment by schiessle** [comment]
@jancborchardt Thanks. As discussed we need three icons:

* not verified
* verification in progress
* verified

**Comment by schiessle** [comment]
@jancborchardt I just pushed the last changes for the code. The strings "verification in progress", "verified", "click to verify" here https://github.com/nextcloud/server/pull/3869/commits/24e1395964f51a83698d970c74534f91993c0f9f#diff-e999cb367e8eac4edfd688b687c55730L114 need to be replaced with nice looking icons. Maybe it make sense to keep the strings as a tooltip.

Email has the status "verification in progress" and "verifed" only and is not clickable because email verification doesn't need any additional user interaction. The lookup server send automatically a verification mail if you set an new email address and we check in the background if the verification was completed.

**Comment by jancborchardt** [comment]
@schiessle what do you think about these icons:
verification
1. Click to verify
2. Verification in progress
3. Verified
4. and 5: Error state – which one do you prefer?

**Comment by schiessle** [comment]
@jancborchardt looks nice. I'm not sure if we need a error state. My plan was to just go back from "verify in progress" to "not verified" and then the user can start again. Can you commit the icons? Thanks!

**Comment by jancborchardt** [comment]
@schiessle I added the icons and also would say we simplify the wording to:
- Verify (cause »click« isn’t on mobile ;)
- Verifying …
- Verified

Can you make all the changes for that, let me know, and then I fix the last layout issues?

**Comment by schiessle** [comment]
@jancborchardt verification buttons are in and they look really nice! 😃 

Can you have a look at the layout, there are still some glitches on different window sizes. Thanks!

File: apps/files_sharing/lib/Controller/ShareesAPIController.php [hunk 5]
```diff
@@ -654,13 +654,14 @@ protected function getEmail($search) {
 
 	protected function getLookup($search) {
 		$isEnabled = $this->config->getAppValue('files_sharing', 'lookupServerEnabled', 'no');
+		$lookupServerUrl = $this->config->getSystemValue('lookup_server', 'https://lookup.nextcloud.com');
```
Comment (nickvergessen) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: Please `rtrim('/')` because people will mostlikely add a trailing slash, which might or might not cause problems...

File: lib/private/Accounts/AccountManager.php [hunk 6]
```diff
@@ -171,21 +284,25 @@ protected function buildDefaultUserRecord(IUser $user) {
 				[
 					'value' => $user->getDisplayName(),
 					'scope' => self::VISIBILITY_CONTACTS_ONLY,
+					'verified' => '0',
```
Comment (nickvergessen) [inline lib/private/Accounts/AccountManager.php]: constant?

File: settings/BackgroundJobs/VerifyUserData.php [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+	public function __construct(AccountManager $accountManager = null,
+								IUserManager $userManager = null,
+								IClientService $clientService = null,
+								ILogger $logger = null,
+								IConfig $config = null
+	) {
+		$this->accountManager = $accountManager !== null ? $accountManager : \OC::$server->query(AccountManager::class);
+		$this->userManager = $userManager !== null ? $userManager : \OC::$server->getUserManager();
+		$this->httpClientService = $clientService !== null ? $clientService : \OC::$server->getHTTPClientService();
+		$this->logger = $logger !== null ? $logger : \OC::$server->getLogger();
```
Comment (nickvergessen) [inline settings/BackgroundJobs/VerifyUserData.php]: kill all this, DI for background jobs is working fine nowerdays

File: settings/BackgroundJobs/VerifyUserData.php [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+
+		if ($config !== null) {
+			$this->lookupServerUrl = $config->getSystemValue('lookup_server', 'https://lookup.nextcloud.com');
```
Comment (nickvergessen) [inline settings/BackgroundJobs/VerifyUserData.php]: rtrim

File: settings/BackgroundJobs/VerifyUserData.php [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+		} else {
+			$this->lookupServerUrl = \OC::$server->getConfig()->getSystemValue('lookup_server', 'https://lookup.nextcloud.com');
```
Comment (nickvergessen) [inline settings/BackgroundJobs/VerifyUserData.php]: only use the DI way

File: settings/templates/personal.php [hunk 10]
```diff
@@ -8,6 +8,18 @@
...
 ?>
 
+<div id="verification-dialog">
+
+	<!-- dialog used to verify personal information such as the users website, email address, etc -->
+	<div class="verification-dialog-content">
+		<button id="verify-dialog-close" class="icon-close svg"></button>
+		<span class="explainVerification">How to verify your account details</span><br />
+		<span class="verificationCode">verification code</span>
+		<p>It can take up to 24 hours before the account is displayed as verified.</p>
```
Comment (nickvergessen) [inline settings/templates/personal.php]: missing translation same in the 3 lines above

File: settings/templates/personal.php [hunk 11]
```diff
@@ -99,6 +111,20 @@
...
+				<img id="verify-email" <?php
+				switch($_['emailVerification']) {
+					case \OC\Accounts\AccountManager::VERIFICATION_IN_PROGRESS:
+						print_unescaped('src="' . image_path('core', 'actions/verifying.svg') . '" title="Verifying..."');
+						break;
+					case \OC\Accounts\AccountManager::VERIFIED:
+						print_unescaped('src="' . image_path('core', 'actions/verified.svg') . '" title="Verified"');
+						break;
+					default:
+						print_unescaped('src="' . image_path('core', 'actions/verify.svg') . '" title="Verify"');
```
Comment (nickvergessen) [inline settings/templates/personal.php]: title not translated, same above

File: settings/templates/personal.php [hunk 12]
```diff
@@ -151,6 +177,20 @@
 				<label for="website"><?php p($l->t('Website')); ?></label>
 				<span class="icon-password"/>
 			</h2>
+			<span class="verify">
+				<img id="verify-website" <?php
+				switch($_['websiteVerification']) {
+					case \OC\Accounts\AccountManager::VERIFICATION_IN_PROGRESS:
+						print_unescaped('src="' . image_path('core', 'actions/verifying.svg') . '" title="Verifying..."');
```
Comment (nickvergessen) [inline settings/templates/personal.php]: Use `…` and translation

File: settings/templates/personal.php [hunk 13]
```diff
@@ -164,6 +204,19 @@
 				<label for="twitter"><?php p($l->t('Twitter')); ?></label>
 				<span class="icon-password"/>
 			</h2>
+			<span class="verify">
+				<img id="verify-twitter" <?php
+				switch($_['twitterVerification']) {
+					case \OC\Accounts\AccountManager::VERIFICATION_IN_PROGRESS:
+						print_unescaped('src="' . image_path('core', 'actions/verifying.svg') . '" title="Verifying..."');
```
Comment (nickvergessen) [inline settings/templates/personal.php]: ... + t

**Review by nickvergessen — CHANGES_REQUESTED** [review]

**Comment by jancborchardt** [comment]
@schiessle are you going to rebase before? Then I’ll wait with checking the layout

**Comment by schiessle** [comment]
@jancborchardt rebased, please have a look. Thanks.

**Comment by jancborchardt** [comment]
Fixed the layout to use the `.popovermenu` as we do for the file actions and the federationScopeMenu too. Here’s a mockup how it will look like when shown via .removeClass('hidden') and positioned with top/right which needs to be done via JS:
capture du 2017-04-26 13-19-15

@schiessle can you adjust the code to use a `class="verification-dialog"` instead of the ID, and like with the federationScopeMenu I think we need to place the dialog directly next to each verification icon.
Basically it should work like we already did in the federationScopeMenu.

**Comment by schiessle** [comment]
@nickvergessen all your comments are addressed, please have a final look code wise

**Comment by schiessle** [comment]
@jancborchardt while your mockup looks great I'm not sure how to make this happen, especially in the remaining few hours. Maybe @ChristophWurst or @rullzer could have a look? I think you also created the other popover menus.

File: settings/templates/personal.php (hunk 10 above)
Comment (schiessle) [inline settings/templates/personal.php]: added it for the "It can take up to 24 hours...", the other two strings are just placeholder and will never be shown to the user

File: settings/css/settings.css [hunk 14]
```diff
@@ -136,14 +136,28 @@ input#openid, input#webdav { width:20em; }
...
 
+/* verify accounts */
 #personal-settings-container .verify {
+	position: relative;
 	padding: 10px;
 	float: right;
 }
-
+#personal-settings-container .verify,
 #personal-settings-container .verify-action {
```
Comment (schiessle) [inline settings/css/settings.css]: @jancborchardt  that's wrong. Only "verify-action" should have a pointer in mouse over. Email verification doesn't has a popover and therefore never should show a mouse pointer. I will fix it. Just want to make sure that we don't change it back and forth. 😉

**Comment by jancborchardt** [comment]
Ok so I switched the code to use individual popovers instead of one main one:
capture du 2017-04-26 14-14-21

All the styling works out of the box. The only changes we need to make:
- [ ] Switch the JS to use the class '.verification-dialog` which is an element right inside the `<div class="verify">`
- [ ] Check if it correctly closes again on clicking outside and opening other menus also, as menus should

cc @nextcloud/javascript

File: apps/lookup_server_connector/lib/UpdateLookupServer.php (hunk 3 above)
Comment (blizzz) [inline apps/lookup_server_connector/lib/UpdateLookupServer.php]: you can still have the default server stored in a const, if you need it in several places, and use it as default parameter. It's a more intuitive and especially consistent.

File: settings/js/personal.js [hunk 15]
```diff
@@ -215,10 +215,12 @@ $(document).ready(function () {
 			OC.generateUrl('/settings/users/{account}/verify', {account: accountId}),
 			{method: 'GET'}
 		).done(function(data) {
-			showVerifyDialog(data.msg, data.code);
+			var dialog = $(this).closest('.verify').children('.verification-dialog');
```
Comment (jancborchardt) [inline settings/js/personal.js]: Maybe this can also be simplified since `.verification-dialog` is the element directly after the `.verify-action` element which this click handler is bound to.

Any idea @nextcloud/javascript @MorrisJobke @ChristophWurst?

**Comment by jancborchardt** [comment]
@schiessle I would also show the verification button for fields when they are not public yet. Verification is always good.

**Comment by jancborchardt** [comment]
@schiessle I made the verification icons also show for private fields (as said above) and fixed the width – please do not make it so wide, it’s just too huge to read and also doesn’t work well on mobile. ;) Now it’s right under the input field and looks nice.

Todo:

- [ ] The verification icon should show directly when there is a value put into the field. Currently done via PHP so it will be only visible after a hard refresh → bad. Any help @nextcloud/javascript?
- [ ] On mobile, the icon is a bit shifted. I’ll look into that.

**Comment by jancborchardt** [comment]
@schiessle I will try to get some JS help to directly show the fields when something has been put into there. ;)

**Comment by schiessle** [comment]
> I will try to get some JS help to directly show the fields when something has been put into there. ;)

Not needed, it is already done! 😃 

But I don't understand this commit: https://github.com/nextcloud/server/pull/3869/commits/47b203d7255e99a028e67c113bb46ab452cff447

It doesn't make any sense to ask the lookup server to verify your email address, twitter handle or webpage if the data is not published to the lookup server.

**Comment by schiessle** [comment]
@jancborchardt please check the verification dialog for the website. There the signature is much longer because we don't have to deal with 140char limit. Not sure how to fit this into the popup

**Comment by jancborchardt** [comment]
@nickvergessen are your concerns addressed? If so can you re-review? :)

**Comment by MorrisJobke** [comment]
The feedback is very cumbersome. You click on the icon and nothing happens. Then after some while (when the request is back) the icon changes to the blue one and you need to click on it a second time to see the steps. Also those steps look really ugly:

bildschirmfoto 2017-04-27 um 00 03 13

This need to be more straightforward. The UX currently is really bad :/ It took me a while to discover this even if I actively search for it.

**Comment by nickvergessen** [comment]
For me the personal page is totally broken:
> nextcloud_-_2017-04-27_10 35 45

**Comment by schiessle** [comment]
@MorrisJobke don't know what happens on your system. Are you sure to have the latest version? For example the text overflow in your screenshot is fixed with https://github.com/nextcloud/server/pull/3869/commits/ea9c4b297d4d25bd5f4344630648bf3b2921a203 for me.

Also I don't understand the workflow you describe. I just need to click once, the popup shows up and the icon changes.

**Comment by schiessle** [comment]
@nickvergessen no idea what's happening on your side. Can anyone reproduce this? I can't...

File: settings/templates/personal.php [hunk 16]
```diff
@@ -99,6 +99,20 @@
 				<label for="email"><?php p($l->t('Email')); ?></label>
 				<span class="icon-password"/>
 			</h2>
+			<div class="verify <?php if ($_['email'] === ''  || $_['emailScope'] !== 'public') p('hidden'); ?>">
+				<img id="verify-email" <?php
+				switch($_['emailVerification']) {
+					case \OC\Accounts\AccountManager::VERIFICATION_IN_PROGRESS:
+						print_unescaped('src="' . image_path('core', 'actions/verifying.svg') . '" title="' . \OC::$server->getL10N()->t('Verifying…') . '"');
```
Comment (nickvergessen) [inline settings/templates/personal.php]: `p($l->t(` no need to get the class all the time again...

Comment (nickvergessen) [inline settings/templates/personal.php]: And thats for all templates

**Review by nickvergessen — CHANGES_REQUESTED** [review]

**Comment by nickvergessen** [comment]
there have been invalid calls to getL10n without the appid, which broke it for me, because I have display_errors enabled...
Fixed with my last commit

**Comment by nickvergessen** [comment]
### Steps
1. Enter a website
2. Make it public
3. Click on the verify thingy, copy the file name
4. Go to your filesystem and create the file
4. Click anywhere (simulating you went to filesystem and just clicked on your browser again)
5. The verify window is closed and you can not open it anymore, so you can not get the content of the file....

**Comment by schiessle** [comment]
@nickvergessen I know... I'm already working on a solution

File: settings/templates/personal.php [hunk 17]
```diff
@@ -103,14 +103,15 @@
...
 						break;
 					case \OC\Accounts\AccountManager::VERIFIED:
-						print_unescaped('src="' . image_path('core', 'actions/verified.svg') . '"' . \OC::$server->getL10N()->t('Verified') . '"');
+						print_unescaped('src="' . image_path('core', 'actions/verified.svg') . '""');
 						break;
 					default:
-						print_unescaped('src="' . image_path('core', 'actions/verify.svg') . '" title="' . \OC::$server->getL10N()->t('Verify') . '"');
+						print_unescaped('src="' . image_path('core', 'actions/verify.svg') . '"');
 				}
+				print_unescaped(' title="' . $_['websiteMessage'] . '""');
```
Comment (nickvergessen) [inline settings/templates/personal.php]: double `""` at the end...
btw ` title="` could also be in the template directly, no need to wrap it with print_unescapced....
Also the message could then use `p(` to avoid problems with translations

File: settings/templates/personal.php [hunk 18]
```diff
@@ -103,14 +103,15 @@
...
-						print_unescaped('src="' . image_path('core', 'actions/verifying.svg') . '" title="' . \OC::$server->getL10N()->t('Verifying …') . '"');
+						print_unescaped('src="' . image_path('core', 'actions/verifying.svg') . '"');
 						break;
 					case \OC\Accounts\AccountManager::VERIFIED:
-						print_unescaped('src="' . image_path('core', 'actions/verified.svg') . '"' . \OC::$server->getL10N()->t('Verified') . '"');
+						print_unescaped('src="' . image_path('core', 'actions/verified.svg') . '""');
 						break;
 					default:
-						print_unescaped('src="' . image_path('core', 'actions/verify.svg') . '" title="' . \OC::$server->getL10N()->t('Verify') . '"');
+						print_unescaped('src="' . image_path('core', 'actions/verify.svg') . '"');
```
Comment (nickvergessen) [inline settings/templates/personal.php]: same here, move `src="` out of the php code and use `p(` for the image

**Review by nickvergessen — DISMISSED** [review]

File: settings/templates/personal.php [hunk 19]
```diff
@@ -151,8 +166,30 @@
...
+					default:
+						print_unescaped('src="' . image_path('core', 'actions/verify.svg') . '" class="verify-action"');
+				}
+				print_unescaped(' title="' . $_['websiteMessage'] . '""');
+				?>>
+				<div class="verification-dialog popovermenu bubble menu">
+					<div class="verification-dialog-content">
+						<p class="explainVerification"></p>
+						<p class="verificationCode"></p>
+						<p><?php p(\OC::$server->getL10N('settings')->t('It can take up to 24 hours before the account is displayed as verified.'));?></p>
```
Comment (nickvergessen) [inline settings/templates/personal.php]: use `$l->t(` instead of creating a new instance of the lang

File: settings/templates/personal.php [hunk 20]
```diff
@@ -164,8 +201,30 @@
...
+					default:
+						print_unescaped('src="' . image_path('core', 'actions/verify.svg') . '" class="verify-action"');
+				}
+				print_unescaped(' title="' . $_['twitterMessage'] . '""');
+				?>>
+				<div class="verification-dialog popovermenu bubble menu">
+					<div class="verification-dialog-content">
+						<p class="explainVerification"></p>
+						<p class="verificationCode"></p>
+						<p><?php p(\OC::$server->getL10N('settings')->t('It can take up to 24 hours before the account is displayed as verified.'));?></p>
```
Comment (nickvergessen) [inline settings/templates/personal.php]: use `$l->t(` instead of creating a new instance of the lang

**Comment by schiessle** [comment]
@nickvergessen you should now be able to re-open the verification dialog

File: settings/BackgroundJobs/VerifyUserData.php [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+	 * verify web page
+	 *
+	 * @param array $argument
+	 * @return bool true if we could check the verification code, otherwise false
+	 */
+	protected function verifyWebsite(array $argument) {
+
+		$result = false;
+
+		$url = rtrim($argument['data'], '/') . '/well-known/' . 'CloudIdVerificationCode.txt';
```
Comment (LukasReschke) [inline settings/BackgroundJobs/VerifyUserData.php]: As per the RFC that should be `.well-known`.

https://tools.ietf.org/html/rfc5785

File: settings/Controller/UsersController.php [hunk 22]
```diff
@@ -493,6 +517,92 @@ public function destroy($id) {
...
+			case 'verify-twitter':
+				$accountData[AccountManager::PROPERTY_TWITTER]['verified'] = AccountManager::VERIFICATION_IN_PROGRESS;
+				$msg = $this->l10n->t('In order to verify your Twitter account post following tweet on Twitter (please make sure to post it without any line breaks):');
+				$code = $codeMd5;
+				$type = AccountManager::PROPERTY_TWITTER;
+				$data = $accountData[AccountManager::PROPERTY_TWITTER]['value'];
+				break;
+			case 'verify-website':
+				$accountData[AccountManager::PROPERTY_WEBSITE]['verified'] = AccountManager::VERIFICATION_IN_PROGRESS;
+				$msg = $this->l10n->t('In order to verify your Website store following content in your web-root at \'CloudIdVerificationCode.txt\' (please make sure that the complete text is in one line):');
```
Comment (LukasReschke) [inline settings/Controller/UsersController.php]: Mhm. The code seems to check against `.well-known/CloudIdVerificationCode.txt` instead.

Comment (schiessle) [inline settings/Controller/UsersController.php]: good catch, will update the text

File: settings/BackgroundJobs/VerifyUserData.php (hunk 21 above)
Comment (schiessle) [inline settings/BackgroundJobs/VerifyUserData.php]: indeed, just a type... will fix

**Comment by schiessle** [comment]
@LukasReschke issues fixed

**Comment by MorrisJobke** [comment]
> Also I don't understand the workflow you describe. I just need to click once, the popup shows up and the icon changes.

Works now, but a loading indicator is still something I would like to see here. If your connection to the server is slow there is no feedback and later the popup appear magically.

File: settings/personal.php [hunk 23]
```diff
@@ -181,6 +185,28 @@
...
+$tmpl->assign('twitterVerification', $userData[\OC\Accounts\AccountManager::PROPERTY_TWITTER]['verified']);
+$tmpl->assign('emailVerification', $userData[\OC\Accounts\AccountManager::PROPERTY_EMAIL]['verified']);
+
+$needVerifyMessage = [\OC\Accounts\AccountManager::PROPERTY_EMAIL, \OC\Accounts\AccountManager::PROPERTY_WEBSITE, \OC\Accounts\AccountManager::PROPERTY_TWITTER];
+
+foreach ($needVerifyMessage as $property) {
+
+	switch ($userData[$property]['verified']) {
+		case \OC\Accounts\AccountManager::VERIFIED:
+			$message = $l->t('Verifying');
```
Comment (LukasReschke) [inline settings/personal.php]: Shouldn't that be "verified"?

**Comment by jancborchardt** [comment]
> Works now, but a loading indicator is still something I would like to see here. If your connection to the server is slow there is no feedback and later the popup appear magically.

Yes, the popup should open immediately. No reason to have that blocked by the server connection cause it’s just info in the popup which will take you time to do anyway.

**Review by LukasReschke — APPROVED** [review]
Good enough for me

**Review by rullzer — APPROVED** [review]
Fine by me for now. Lets do small fixes later!

**Review by rullzer — CHANGES_REQUESTED** [review]
But failing tests....

**Comment by schiessle** [comment]
spinner can be added in the bugfixing / polishing phase. I will look at the failing tests

**Comment by schiessle** [comment]
> Yes, the popup should open immediately. No reason to have that blocked by the server connection cause it’s just info in the popup which will take you time to do anyway.

it updates the account table which triggers a hook which again triggers a update send to the lookup server. At the moment we try once in real time to update the server and if it fails we create a cronjob to retry later. I will look into the possibility to add a spinner once the pr is merged

**Comment by MorrisJobke** [comment]
I rebased to trigger CI - restarting is not possible for the jobs before the drone mess up. :/

**Review by MorrisJobke — APPROVED** [review]
Tested and works:+1:

**Comment by MorrisJobke** [comment]
unit tests run fine -> merge
